### PR TITLE
DTSERWONE-846 - Java SDK - set http connection timeout

### PR DESCRIPTION
**Pull request by grmeyer-hw-dev**
- Support URL connection and read timeout with default value
- Add documentation
- Merge MultipartRequest to Request to have single class to handle request and common HTTP header builder.
- Add unit test to test read timeout

Sample code how to set timeout value:
```java
   ...
   //300 milliseconds
   int connectionTimeout = 300;
   
   // 2 seconds  
   int readTimout = 2000;
   
   Hyperwallet client = new Hyperwallet(username, password, programToken, connectionTimeout, readTimout);
```

### CODE REVIEW

**Review by fmattos-hw — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,12 @@
 Changelog
 =========
+
+2.4.3
+-----------------
+- Added connection and read timeout support
+- Added documentation over Hyperwallet Api client
+- Merged the implantation from Multipart Request to Request
```
Comment (peter-joseph) [inline CHANGELOG.md:8]: typo -> `implementation` instead

File: src/main/java/com/hyperwallet/clientsdk/util/Request.java [hunk 2]
```diff
@@ -1,22 +1,16 @@
...
-import java.io.BufferedReader;
-import java.io.ByteArrayOutputStream;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
-import java.io.OutputStreamWriter;
-import java.lang.reflect.Field;
-import java.lang.reflect.Modifier;
+import cc.protea.util.http.Response;
+import com.hyperwallet.clientsdk.util.Multipart.MultipartData;
```
Comment (peter-joseph) [inline src/main/java/com/hyperwallet/clientsdk/util/Request.java:5]: what's the reason of merging normal request object with `MultipartRequest` which is used for Document/File upload?

File: src/test/java/com/hyperwallet/clientsdk/util/HyperwalletApiClientTest.java [hunk 3]
```diff
@@ -2199,6 +2205,67 @@ public void testPostProxy_withIncorrectURL() {
         }
     }
 
+    @Test
+    public void testGet_exceedReadTimeout() throws InterruptedException {
```
Comment (peter-joseph) [inline src/test/java/com/hyperwallet/clientsdk/util/HyperwalletApiClientTest.java:2209]: is there a test for connection timeout, simulation?

File: src/main/java/com/hyperwallet/clientsdk/Hyperwallet.java [hunk 4]
```diff
@@ -30,12 +32,45 @@ public class Hyperwallet {
...
      * @param server                API server url
      * @param hyperwalletEncryption API encryption data
+     * @param connectionTimeout     the timeout value that will be used for making new connections to the Hyperwallet API (in milliseconds).
+     * @param readTimeout           the timeout value that will be used when reading data from an established connection to
+     *                              the Hyperwallet API (in milliseconds).
      */
     public Hyperwallet(final String username, final String password, final String programToken, final String server,
-            final HyperwalletEncryption hyperwalletEncryption) {
-        apiClient = new HyperwalletApiClient(username, password, VERSION, hyperwalletEncryption);
+            final HyperwalletEncryption hyperwalletEncryption, final int connectionTimeout, final int readTimeout) {
```
[the remark this answers is not in the document]
Comment (grmeyer-hw-dev) [reply src/main/java/com/hyperwallet/clientsdk/Hyperwallet.java:40]: Zero means infinite timeout based on the Java docs:
https://docs.oracle.com/javase/8/docs/api/java/net/URLConnection.html#setConnectTimeout-int-
https://docs.oracle.com/javase/8/docs/api/java/net/URLConnection.html#setReadTimeout-int-
```
public void setConnectTimeout(int timeout)
Sets a specified timeout value, in milliseconds, to be used when opening a communications link to the resource referenced by this URLConnection. If the timeout expires before the connection can be established, a java.net.SocketTimeoutException is raised. A timeout of zero is interpreted as an infinite timeout.
```

File: src/main/java/com/hyperwallet/clientsdk/util/Request.java (hunk 2 above)
Comment (grmeyer-hw-dev) [reply src/main/java/com/hyperwallet/clientsdk/util/Request.java:5]: Good question, I'm thinking in move this change to a new PR, but on introducing timeout I saw most of the MultipartRequest is duplicated from Request. The only missing part for Request is writing the multipart into the connection output steam. 

Another benefit is having a single method to prepare the request with HTTP hearders and connections timeout.

File: CHANGELOG.md (hunk 1 above)
Comment (grmeyer-hw-dev) [reply CHANGELOG.md]: good catch

File: src/test/java/com/hyperwallet/clientsdk/util/HyperwalletApiClientTest.java (hunk 3 above)
Comment (grmeyer-hw-dev) [reply src/test/java/com/hyperwallet/clientsdk/util/HyperwalletApiClientTest.java:2209]: Good call. I'm still investigate how to test Socket timeout 
Do you know any approach ?

File: src/main/java/com/hyperwallet/clientsdk/util/Request.java (hunk 2 above)
Comment (peter-joseph) [reply src/main/java/com/hyperwallet/clientsdk/util/Request.java:5]: Oh is this better implemented with Inheritance?, e.g. `MultipartRequest` <extends> `Request`

File: src/test/java/com/hyperwallet/clientsdk/util/HyperwalletApiClientTest.java (hunk 3 above)
Comment (peter-joseph) [reply src/test/java/com/hyperwallet/clientsdk/util/HyperwalletApiClientTest.java:2209]: you can target this service https://httpstat.us/

File: src/main/java/com/hyperwallet/clientsdk/util/MultipartRequest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * The Constructor takes the url as a String, a proxy as a Proxy, and proxy credentials as a String.
+     *
+     * @param url               The url parameter does not need the query string parameters if they are going to be supplied via calls to
+     *                          {@link #addQueryParameter(String, String)}. You can, however, supply the query parameters in the URL if you wish.
+     * @param connectionTimeout A specified timeout value, in milliseconds, to establish communications link to the resource by
+     *                          {@link  java.net.URLConnection}.
+     * @param readTimeout       A specified timeout, in milliseconds, for reading data from an established connection to the resource
+     *                          by {@link  java.net.URLConnection}.
```
Comment (peter-joseph) [inline src/main/java/com/hyperwallet/clientsdk/util/MultipartRequest.java:32]: 👍🏽

File: src/main/java/com/hyperwallet/clientsdk/util/MultipartRequest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    public static final String CRLF = "\r\n";
+    private static final String SEPARATOR = "--";
+
+    /**
+     * The Constructor takes the url as a String, a proxy as a Proxy, and proxy credentials as a String.
+     *
+     * @param url               The url parameter does not need the query string parameters if they are going to be supplied via calls to
+     *                          {@link #addQueryParameter(String, String)}. You can, however, supply the query parameters in the URL if you wish.
+     * @param connectionTimeout A specified timeout value, in milliseconds, to establish communications link to the resource by
+     *                          {@link  java.net.URLConnection}.
```
Comment (peter-joseph) [inline src/main/java/com/hyperwallet/clientsdk/util/MultipartRequest.java:30]: 👍🏽

File: src/main/java/com/hyperwallet/clientsdk/util/MultipartRequest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import java.io.IOException;
+import java.net.Proxy;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Represents the HTTP Multipart Request message.
+ */
+public class MultipartRequest extends Request{
```
Comment (peter-joseph) [inline src/main/java/com/hyperwallet/clientsdk/util/MultipartRequest.java:18]: 👍🏽

**Review by peter-joseph — APPROVED** [review]
LGTMN

File: src/main/java/com/hyperwallet/clientsdk/Hyperwallet.java [hunk 8]
```diff
@@ -30,12 +32,45 @@ public class Hyperwallet {
...
+    /**
+     * Create Hyperwallet SDK instance
+     *
+     * @param username              API key assigned
+     * @param password              API Password assigned
+     * @param programToken          API program token
+     * @param server                API server url
+     * @param hyperwalletEncryption API encryption data
+     */
+    public Hyperwallet(final String username, final String password, final String programToken, final String server,
```
[the remark this answers is not in the document]
Comment (grmeyer-hw-dev) [reply src/main/java/com/hyperwallet/clientsdk/Hyperwallet.java:55]: Yes, most of the usage is on this constructor.

**Review by fzhang4-hw — APPROVED** [review]

**Comment by fzhang4-hw** [comment]
I see some duplicate git commits. Maybe we can squash and merge

**Pull request merged**